#187300141 User email verification

musabehonore opened the pull request:
# What does this PR do?
This PR implements the functionality for users verify their accounts using their emails.

# Description of Task to be completed?
A guest should be able to receive a verification email to verify their account after he/she successfully registers ( signs up ).

# How should this be manually tested?
### For testing only
npm install
npm run test
### For running the app
npm run install
npm run demigrate
npm run migrate
npm run seed
npm run dev

# Any background context you want to provide?
This feature requires to use a service called Nodemailer , we just send an email to the registered person containing a jwt token that we will verify and then mark the user **verified**.

# What are the relevant pivotal tracker/Trello stories?
-Story ID: #187300141

fab-ryan commented on src/controllers/email.ts:
```diff
@@ -0,0 +1,32 @@
+import nodemailer from 'nodemailer';
+import dotenv from 'dotenv';
+
+dotenv.config();
+
+/**
+ * Sends an email using nodemailer.
+ * @returns {Promise<void>} A promise that resolves when the email is sent.
+ */
+export async function sendEmail() {
+  const transporter = nodemailer.createTransport({
+    host: 'smtp.gmail.com',
+    port: 465,
+    secure: true,
+    auth: {
+      user: process.env.USER,
+      pass: process.env.PASS,
+    },
+  });
+
+  const info = await transporter.sendMail({
+    from: '"ALPHA Atlp Cohort 29" <musabemusabe10@gmail.com>',
+    to: 'musabehonore@gmail.com',
+    subject: 'Your email verification',
+    html: `
+    <h1>Y'ello there You just signed up somewhere ....</h1>
+
```
wrong implementation create File like in utils that handles everything and create html/css file that have good format of email that represent every.

musabehonore replied:
Yeah

fab-ryan commented on src/controllers/loginController.ts:
```diff
@@ -57,6 +58,29 @@ const validatingUser = async (req: Request, res: Response) => {
+
+      // Send verification email if the User is not verified yet
+
+      const verificationToken = jwt.sign({ email: user.email }, secret, {
+        expiresIn: '15m',
+      });
+
+      user.verificationToken = verificationToken;
+      await user.save();
+      const { email } = user;
+      const { name } = user;
+
+      const mailOptions = {
+        to: email,
+        subject: 'Your Email verification',
+        template: 'email',
+        context: {
+          name,
+          verificationLink: `http://localhost:3000/api/users/verify-email/${verificationToken}`,
+        },
```
make this varification link dynamic because in production we will not have this

fab-ryan replied:
this can happened when user didn't verfiy his/her info and no loggin should happened

fab-ryan commented on src/controllers/loginController.ts:
```diff
@@ -57,6 +58,29 @@ const validatingUser = async (req: Request, res: Response) => {
+      await sendEmail(mailOptions);
+
       // generate success Response
       return sendResponse<string>(res, 200, token, 'Logged In Successfully');
```
don't return token let them go to check in  email

fab-ryan commented on src/controllers/userController.ts:
```diff
@@ -48,6 +50,26 @@ export const createUser = async (
+      template: 'email',
+      context: {
+        name,
+        verificationLink: `http://localhost:3000/api/users/verify-email/${verificationToken}`,
```
this links also should be dynamic

fab-ryan commented on src/controllers/userVerifyController.ts:
```diff
@@ -0,0 +1,34 @@
+  const { secret } = getDatabaseConfig();
+
+  try {
+    const decoded = jwt.verify(token, secret) as { email: string };
```
create function that hanlde verify and assgn token please

fab-ryan commented on src/controllers/userVerifyController.ts:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    user.verified = true;
+    user.verificationToken = null;
```
why are you storing verification token please we don't want that

fab-ryan commented on src/database/migrations/20240325172706-create-user.js:
```diff
@@ -36,6 +36,15 @@ module.exports = {
         allowNull: false,
         type: Sequelize.DATE,
       },
+      verified: {
+        allowNull: false,
+        type: Sequelize.BOOLEAN,
+        defaultValue: false,
+      },
+      verificationToken: {
+        allowNull: true,
+        type: Sequelize.STRING,
+      },
```
wrong implementation please create new migration that add those field or column

fab-ryan replied:
create different fiile please

fab-ryan reviewed (changes requested):
Work on those comments

fab-ryan commented on src/controllers/userController.ts:
```diff
@@ -48,6 +50,26 @@ export const createUser = async (
+    };
+    await sendEmail(mailOptions);
+
     return sendResponse<string>(res, 201, token, 'User created successfully!');
```
data:{
message:"verigy your account in:"}
}

fab-ryan reviewed (changes requested):
correct those mistakes

fab-ryan commented on src/utils/email.ts:
```diff
@@ -0,0 +1,50 @@
+    port: 465,
+    secure: true,
+    auth: {
+      user: process.env.USER,
+      pass: process.env.PASS,
```
please name well enviromnent

fab-ryan reviewed (changes requested):
Please correct that

fab-ryan commented on src/database/migrations/20240410100649-add-verified-to-user.js:
```diff
@@ -0,0 +1,14 @@
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+  async up(queryInterface, Sequelize) {
+    await queryInterface.addColumn('users', 'verifiedAccount', {
```
we already have this verified use it

musabehonore replied:
The verified we have is the Google verified, so we need another verified that verifies account on our platform (excluding google).

That is why we use another field **verifiedAccount**

fab-ryan replied:
please remove this it's was already added by nadine

fab-ryan commented on src/controllers/loginController.ts:
```diff
@@ -1,26 +1,52 @@
+        'User not found or Invalid Credentials'
+      );
     }
+    if (!user.verifiedAccount) {
+      const verificationToken = signToken({ email: user.email }, '15m');
```
please use verify

musabehonore replied:
And about verify function, I have a file in utils containing function , one used to sign a token, and another used to verify a token. So that we don't need to be calling jwt everytime in the logins or elsewhere.

So yeah, I use verify

fab-ryan replied:
no that's what I meant I meant the column use verify it's was added by @Nadinefiona

fab-ryan reviewed (changes requested):
work on those comments

fab-ryan commented on src/database/models/user.ts:
```diff
@@ -96,6 +99,7 @@ export class User extends Model<UserAttributes, UserCreationAttributes> {
       role: this.role,
       createdAt: this.createdAt,
       updatedAt: this.updatedAt,
+      verifiedAccount: this.verifiedAccount,
```
same as here

fab-ryan replied:
go on line 98 you will find verify column

fab-ryan commented on src/database/models/user.ts:
```diff
@@ -118,6 +122,10 @@ export const UserModel = (sequelize: Sequelize) => {
       verified: DataTypes.BOOLEAN,
       createdAt: DataTypes.DATE,
       updatedAt: DataTypes.DATE,
+      verifiedAccount: {
+        type: DataTypes.BOOLEAN,
+        defaultValue: false,
+      },
```
this is not required

fab-ryan commented on src/database/seeders/20240411132319-add-user.js:
```diff
@@ -0,0 +1,55 @@
+          email: 'test3@example.com',
+          phone: '0783556915',
+          address: 'New York USA',
+          verifiedAccount: true,
```
same as this

fab-ryan reviewed (changes requested):
you didn't work on what I gave you

musabehonore commented:
> you didn't work on what I gave you

Ok I understand the suggestion,  (to use the same **verified** field as the one used by Nadine for **google** accounts). I will do it , even though I think there might be crushes in the app as google and account verifications share the same column/field **verified**

musabehonore commented:
> you didn't work on what I gave you

Now done

fab-ryan reviewed (approved):
oky good

musabehonore committed:
feat(email verification): verify email

verify account using a link sent to the user email
use nodemailer to send the emails
users can not login before verifying their email
document the endpoint
test the endpoints

[#187300141]
